xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,256 @@
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	return childrenUpdates, nil
+}
+
+// deltaInClusterUpdate determines whether there was a delta in the cluster update "fields" (meaning irrespective of children)
+// received from the xdsclient vs. the clusterUpdate that was currently present for a given cluster node. This will be used to
+// help decide whether to ping the cluster handler at the end of the handleResp() callback or not.
+func deltaInClusterUpdateFields(clusterUpdateReceived xdsclient.ClusterUpdate, clusterUpdateCurrent xdsclient.ClusterUpdate) bool {
+	if clusterUpdateReceived.ClusterName != clusterUpdateCurrent.ClusterName {
+		return true
+	}
+	if clusterUpdateReceived.EnableLRS != clusterUpdateCurrent.EnableLRS {
+		return true
+	}
+	if clusterUpdateReceived.ClusterType != clusterUpdateCurrent.ClusterType {
+		return true
+	}
+	if clusterUpdateReceived.SecurityCfg != clusterUpdateCurrent.SecurityCfg {
+		return true
+	}
+	if clusterUpdateReceived.MaxRequests != clusterUpdateCurrent.MaxRequests {
+		return true
+	}
+	return false
+}
+
+// handleResp handles a xds response for a particular cluster. This function also handles any logic with regards to any
+// child state that may have changed. At the end of the handleResp(), the clusterUpdate will be pinged in certain situations
+// to try and construct an update to send back to CDS.
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	defer c.clusterHandler.clusterMutex.Unlock()
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		// For a ClusterUpdate, the only update CDS cares about is the most recent one, so opportunistically drain the
+		// update channel before sending the new update.
+		select {
+		case <-c.clusterHandler.updateChannel:
+		default:
+		}
+		c.clusterHandler.updateChannel <- clusterHandlerUpdate{chu: nil, err: err}
+		return
+	}
+
+	// deletedChild helps determine whether this callback will ping the overall clusterHandler to try and construct an
+	// update to send back to CDS. This will be determined by whether there would be a change in the overall clusterUpdate
+	// for the whole tree (ex. change in clusterUpdate for current cluster or a deleted child) and also if there's even
+	// a possibility for the update to build (ex. if a child is created and a watch is started, that child hasn't received
+	// an update yet due to the mutex lock on this callback).
+	var deletedChild bool
+	// deltaInChildUpdateFields determines whether there was a delta in the clusterUpdate fields (forgetting the children).
+	// This will be used to help determine whether to pingClusterHandler at the end of this callback or not.
+	deltaInChildUpdateFields := deltaInClusterUpdateFields(clusterUpdate, c.clusterUpdate)
```

Comment:
Since the processing of aggregated cluster and leaf cluster is so different, how about
```go
c.receivedUpdate = true
oldUpdate := c.clusterUpdate
c.clusterUpdate = clusterUpdate

if not aggregated {
  if deltaInClusterUpdateFields(c.clusterUpdate, oldUpdate)
    send update
  return
}
// aggregated
```

Reply:
Yeah I was thinking about this too. Good idea.

Reply:
Wait. I remember why I couldn't do this. You can't have it branch off if not aggregated. There will be children who will still need to be deleted, which fits it with what I have now. What you have works, but you'd need to delete the children as well. Would you still like this with a loop like for child in child delete child?

Reply:
You know what, I'll just do it your way with a for loop for child in child delete child.

Reply:
Changed to branch and return if not an aggregate cluster.